### PR TITLE
[feat]: add image delete in document cover

### LINKED ISSUE #2076: [FR] Support deleting the cover image inside Images
### Description

* Add a `delete` button for each image.
* Add a `clear all` button on the top-right corner.

Screenshot 2023-03-22 at 14 06 53

### Impact

Users who want to delete the uploaded image.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by GouravShDev**
Fixes #2076 

- Added ChangeCoverPopoverBloc for mantaining state for images path List
- Moved code for fetching previouslyPickedImages inside the bloc as it is app logic

# Result

https://user-images.githubusercontent.com/74348508/227769253-0fbdb90f-2d62-4040-9ab8-96f1e21683f4.mp4

### CODE REVIEW

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover.dart
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover.dart:289]:
```suggestion

      return Container();
```

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart:46]: I would personally prefer this being part of the BLoC as a private field that is initialized. In case there is a need for it in another event, or if the event calling this method is called again and again.

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart:54]:
```suggestion
    final imageNames = prefs.getStringList(kLocalImagesKey) ?? [];
  
    if (imageNames.isEmpty) {
      return [];
    }

    imageNames.removeWhere((name) => !File(name).existsSync());
```
Probably would need some testing, but to me this is cleaner, and also avoids another iteration of imageNames.

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart:60]: Yeah, definitely should just be part of the BLoC, to ensure getInstance is only called once during it's lifetime.

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart
Comment (GouravShDev) [reply frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart:46]: You mean instance of shared preference as private field?

Comment (Xazin) [reply frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart:46]: Yes,

`late final SharedPreferences _prefs;`

Comment (GouravShDev) [reply frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart:46]: I'm having trouble initializing future as field inside bloc. Can you share code for the reference ? Thank you

Comment (Xazin) [reply frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart:46]: The code sample I sent is what you need to do.

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart:12]:
```suggestion
  late final SharedPreferences _prefs;
  final _initCompleter = Completer<void>();

  ChangeCoverPopoverBloc() : super(const ChangeCoverPopoverState.initial()) {
  
    SharedPreferences.getInstance().then((prefs) {
      _prefs = prefs;
      _initCompleter.complete;
    });
```

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart
Comment (GouravShDev) [reply frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart:46]: Thanks @Xazin. I learned something new.

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart:57]:
```suggestion
  Future<List<String>> _getPreviouslyPickedImagePaths() async {
    await _initCompleter.future;
    
    final imageNames = _prefs.getStringList(kLocalImagesKey) ?? [];
    if (imageNames.isEmpty) {
      return imageNames;
    }

    imageNames.removeWhere((name) => !File(name).existsSync());
  }
```

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart:63]:
```suggestion
  Future<void> _updateImagePathsInStorage(List<String> imagePaths) async {
    await _initCompleter.future;
    _prefs.setStringList(kLocalImagesKey, imagePaths);
  }
```

**Comment by Xazin** [comment]
I made new suggestions to show how you can do the final field for preferences, can you test and let me know if it works? :)

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart
Comment (GouravShDev) [reply frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart:46]: I have a doubt. The code is working even without using _initCompleter. Is it necessary?

Comment (Xazin) [reply frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart:46]: Well, it's primarily to stop an edge case to happen, if let's say for some reason the platform method that returns the SharedPreferences is blocked for some reason, and you instantiate the BLoC with an event, a method could be called where _prefs is not instantiated, it's a rare edge case, but it can happen.

**Comment by Xazin** [comment]
Commit message linter fails, do a quick:
```
git commit --amend -m "fix: amend according to review comments"
git push --force
```

**Comment by Xazin** [comment]
@LucasXu0 All yours

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover_bloc.dart:18]: _initCompleter.complete();

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover.dart
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/cover/change_cover_popover.dart:263]: I think it is more appropriate for hover to display the delete button on the image.

**Comment by GouravShDev** [comment]
@hyj1204 just want to clarify that I have to change color surface and selector color in theme. As this might change color scheme for other parts of the app too. 
And you mentioned:
"You can change current selectorcolor to 1A202C in dark mode."
"You can change current surfacecolor to 1A202C in dark mode."
Should I change both color to same value? This causing no color change on hover

image

**Comment by hyj1204** [comment]
> 
Yes, that's correct. 
Don't worry, I'm currently updating all the colors for the dark mode and the ones I provided to you are the most recent version. You can go ahead and use them. I may make some minor adjustments later on when I'm ready to merge all the dark mode colors together.

**Comment by LucasXu0** [comment]
@GouravShDev Can you rebase or merge the latest branch to resolve the conflicts?

**Comment by GouravShDev** [comment]
@LucasXu0 sure. But there was an issue with hover color on clear all button in dark mode so don't merge it yet

**Comment by Xazin** [comment]
> @LucasXu0 sure. But there was an issue with hover color on clear all button in dark mode so don't merge it yet

You can mark a PR as Draft until it's ready for merging. 👍

**Review by hyj1204 — APPROVED** [review]
I added the updated fill color in the 'Clear All' button.  It looks good to me.

https://user-images.githubusercontent.com/14248245/228895739-c058a3b8-f16e-4520-bb22-568f2bed1100.mov

**Comment by GouravShDev** [comment]
@LucasXu0 the issue with hover color is fixed. Let me know if there are any more changes that needed to be done.

**Comment by LucasXu0** [comment]
Hey, @GouravShDev. There's one more issue I just found. If we delete the image that is being used, the app will throw an error because the image file doesn't exist. I think we should just revert to the built-in images or colors if the selected image is missing.

Screenshot 2023-04-03 at 11 07 39

**Comment by LucasXu0** [comment]
I will merge this pull request first and then create a new issue to follow up on the previously mentioned issue.

**Comment by GouravShDev** [comment]
@LucasXu0 I have created the issue.

**Comment by annieappflowy** [comment]
Should it go back to the default (no cover image) state after I delete the cover image?

**Pull request merged**
